torch/csrc/jit/python/python_ivalue.h

```diff
@@ -42,6 +42,19 @@ struct C10_EXPORT ConcretePyObjectHolder final : PyObjectHolder {
     return py::str(py_obj_);
   }
 
+  std::vector<at::Tensor> extractTensors() override {
+    pybind11::gil_scoped_acquire ag;
+    // We could implement this entirely in C++ via pybind11 but it turns out to
+    // be substantially slower. Namely, the total time taken by markCompleted on
+    // a CUDAFuture is 21.5us with this implementation, but goes up to 58.7us
+    // when using C++. The reason is unclear.
```

Comment:
It might be due to more frequent C++ - Python context switches?

Reply:
That could be one explanation. Another possibility could be that Python byte-compiles and somewhat optimizes the pure Python code? It sounds unlikely though...

I guess that something is going on here beyond what I'm currently understanding, and we might want to investigate, but I also feel we'd be going down a rabbit hole...